contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryBase2_3.sol

```diff
@@ -669,13 +673,29 @@ abstract contract AutomationRegistryBase2_3 is ConfirmedOwner {
     uint256 gasPaymentHexaicosaUSD = (gasWei *
       (paymentParams.gasLimit + paymentParams.gasOverhead) +
       paymentParams.l1CostWei) * paymentParams.nativeUSD; // gasPaymentHexaicosaUSD has an extra 8 zeros because of decimals on nativeUSD feed
-    receipt.gasCharge = SafeCast.toUint96(gasPaymentHexaicosaUSD / paymentParams.billingTokenParams.priceUSD); // has units of attoBillingToken, or "wei"
+
+    uint96 gasCharge18Decimals = SafeCast.toUint96(gasPaymentHexaicosaUSD / paymentParams.billingTokenParams.priceUSD); // has units of attoBillingToken, or "wei"
+    receipt.gasCharge = gasCharge18Decimals;
+    if (decimals < 18) {
+      receipt.gasCharge = SafeCast.toUint96(gasCharge18Decimals / (10 ** (18 - decimals)));
+    } else if (decimals > 18) {
+      receipt.gasCharge = SafeCast.toUint96(gasCharge18Decimals * (10 ** (decimals - 18)));
+    }
+
     receipt.gasReimbursementJuels = SafeCast.toUint96(gasPaymentHexaicosaUSD / paymentParams.linkUSD);
     uint256 flatFeeHexaicosaUSD = uint256(paymentParams.billingTokenParams.flatFeeMilliCents) * 1e21; // 1e13 for milliCents to attoUSD and 1e8 for attoUSD to hexaicosaUSD
     uint256 premiumHexaicosaUSD = ((((gasWei * paymentParams.gasLimit) + paymentParams.l1CostWei) *
       paymentParams.billingTokenParams.gasFeePPB *
       paymentParams.nativeUSD) / 1e9) + flatFeeHexaicosaUSD;
-    receipt.premium = SafeCast.toUint96(premiumHexaicosaUSD / paymentParams.billingTokenParams.priceUSD);
+
+    uint96 premium18Decimals = SafeCast.toUint96(premiumHexaicosaUSD / paymentParams.billingTokenParams.priceUSD); // has units of attoBillingToken, or "wei"
+    receipt.premium = premium18Decimals;
+    if (decimals < 18) {
+      receipt.premium = SafeCast.toUint96(premium18Decimals / (10 ** (18 - decimals)));
+    } else if (decimals > 18) {
+      receipt.premium = SafeCast.toUint96(premium18Decimals * (10 ** (decimals - 18)));
```

Comment:
because we're doing multiplication (line 696) _after_ division (line 691) I think we're potentially losing some precision. Could we instead do something like this...

```sol
uint256 numeratorScalingFactor = decimals > 18 ? 10 ** (decimals - 18) : 1;
uint256 denominatorScalingFactor = decimals < 18 ? 10 ** (18 - decimals) : 1;
receipt.premium = SafeCast.toUint96(
  (premiumHexaicosaUSD * numeratorScalingFactor) / (paymentParams.billingTokenParams.priceUSD * denominatorScalingFactor )
);
```

wdyt?